### PR TITLE
fix(ledgerstate): bound the imported opcert counter

### LINKED ISSUE #4018: ledgerstate: the Mithril import path writes opcert counters without the persistable bound
Follow-up to #3991 / #4001. Two non-blocking review findings on that PR are actionable at the current gouroboros pin.

### 1. `importOpCertCounters` bypasses the persistable bound

`ledger/eras/opcert.go` gates writes to `pool_opcert_sequence` on `MaxPersistableOpCertCounter`, but `importOpCertCounters` (`ledgerstate/import.go:2295`) is a write path into that table that does not call the check.

Its counters come from `decodeOpCertCounters` (`ledgerstate/snapshot.go:1009`), which decodes the certified Mithril counter map at full `uint64` width independently of the gouroboros pin. Unlike the block-apply and forging paths, this one is reachable at v0.202.9.

A counter above the bound therefore reaches `sqlstore.checkedInt64` and fails with `unsigned SQL value %d exceeds int64` — the width-anonymous failure #4001 removed everywhere else. The import aborts atomically so nothing is part-applied, but the class #4001 describes is not complete while this path differs.

### 2. `opcert_persistable_test.go` asserts the constant against itself

`ledger/eras/opcert_persistable_test.go:76` checks `MaxPersistableOpCertCounter` against its own definition (`uint64(math.MaxInt64)` at `ledger/eras/opcert.go:89`), so it cannot detect the drift its name describes — it stays green if either side moves.

The authority for the bound is `sqlstore.checkedInt64`, which `ledger/eras` cannot import. A test placed next to `checkedInt64` asserting it accepts `eras.MaxPersistableOpCertCounter` and rejects that value plus one would actually tie the two together.

### PR DESCRIPTION
**Pull request by wolf31o2**
Fixes #4018

## The write path

`decodeOpCertCounters` decodes the certified HeaderState counter map at the reference's full `uint64` width, independently of the header types the gouroboros pin declares. `importOpCertCounters` therefore could hand `UpdatePoolOpCertSequence` a counter above `eras.MaxPersistableOpCertCounter`, aborting the import at `checkedInt64` with `unsigned SQL value 9223372036854775808 exceeds int64` — the width-anonymous message #4001 replaced on the block-apply and forging paths. The check now runs before the write and names the bound.

Every write path into `pool_opcert_sequence`:

| Path | Bound |
| --- | --- |
| `sqlstore.Store.UpdatePoolOpCertSequence` (`database/plugin/metadata/sqlstore/pool.go`) | the only INSERT into the table |
| `ledger.ledgerProcessBlock` (`ledger/state.go`) | `ValidateOpCertPersistableCounter` before transactions are processed |
| `ledgerstate.importOpCertCounters` (`ledgerstate/import.go`) | added here |
| `sqlstore` rollback DELETE (`pool.go`) | removes rows only |

`ledger/forging/keys.go` and `ledger/forging/builder.go` apply the same bound before the node forges a block whose apply would write.

The denormalized `pool.latest_op_cert_sequence`: the rollback recompute derives it from `MAX(sequence)` over rows already inside the bound; `ImportPool` writes `models.Pool.LatestOpCertSequence`, which is never assigned anywhere and is always zero; the pool registration and retirement certificate inserts write the literal `0`.

## The bound test

`TestMaxPersistableOpCertCounterMatchesSignedColumn` compared `MaxPersistableOpCertCounter` to `uint64(math.MaxInt64)`, its own definition, so it could not detect drift on the store side. Narrowing `checkedInt64` to `math.MaxInt32` leaves the whole `ledger/eras` suite green.

The authority is `sqlstore.checkedInt64`, which is unexported, and `internal/architecture/import_boundary_test.go` forbids anything under `database/` from importing `ledger/`, test files included — so the assertion cannot live next to it. `ledgerstate.TestPersistableOpCertBoundMatchesStore` drives the constant through `UpdatePoolOpCertSequence` in both directions instead: it owns the write path the bound was missing from and already imports `ledger/eras`.

## Related

- #4019 covers the two findings on #4001 that stay blocked on the gouroboros widening; this one is reachable at the current pin because the Mithril counter map is decoded independently of the header types.

---
## Summary by cubic
Fixes #4018. Mithril imports now reject certified opcert counters above `eras.MaxPersistableOpCertCounter` before writing, replacing the generic `checkedInt64` failure with an error that names the supported bound.

- Counters at the bound still import successfully; higher values leave no sequence row.
- Replaces the tautological bound test with a real store test that verifies acceptance at the limit and rejection above it.
- Documents the signed counter range and import validation, and stabilizes a cross-platform Mithril retry test.
- No migration or rollout action is needed because existing rows were already constrained by `checkedInt64`.

<sup>Written for commit 7fee0bface235e7ed36108a40afd573502eae3d7. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/dingo/pull/4030?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Bug Fixes**
  - Snapshot restoration now validates operational certificate counters before saving them.
  - Counters exceeding the supported limit are rejected with a clear error, and failed imports leave no partial data.
  - Counters at the supported maximum are imported successfully.

- **Documentation**
  - Updated architecture and database documentation to describe operational certificate counter imports, validation limits, and rollback behavior.

- **Tests**
  - Added coverage for boundary values, rejected imports, transaction rollback, and consistency between ledger and storage limits.

### CODE REVIEW

File: ledger/eras/opcert_persistable_test.go [hunk 1]
```diff
@@ -70,16 +70,13 @@ func TestValidateOpCertPersistableCounter(t *testing.T) {
...
-		eras.MaxPersistableOpCertCounter,
-	)
-}
+// The bound is the largest value a signed 64-bit SQL column holds, which is
+// what sqlstore.checkedInt64 enforces at the metadata store. That coupling is
+// asserted by ledgerstate.TestPersistableOpCertBoundMatchesStore, which drives
+// the constant through the real store write in both directions. It cannot be
+// asserted here: checkedInt64 is unexported, ledger/eras sits below the
+// database packages, and comparing the constant against its own definition
+// would stay green if either side moved.
```
Comment (chrisguiney) [inline ledger/eras/opcert_persistable_test.go]: Non-blocking, but this sentence overstates what the deleted test missed. `TestMaxPersistableOpCertCounterMatchesSignedColumn` asserted against the `math.MaxInt64` literal and the constant is `uint64(math.MaxInt64)`, so a change to the constant would have failed it. What it could not see is store-side drift: narrowing `checkedInt64` leaves the whole `ledger/eras` suite green, which is exactly the gap `ledgerstate.TestPersistableOpCertBoundMatchesStore` closes. The same wording is in `ledgerstate/opcert_persistable_bound_test.go` ("it stays green if either side moves"), so a correction should cover both files. Same finding as cubic's P3.

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -4510,9 +4510,20 @@ counter is recorded for every applied block -- naming the bound, rather than
...
+not then apply. `ledgerstate.importOpCertCounters` refuses it too: the
+certified HeaderState counter map is decoded at the reference's full
+`uint64` width (`decodeOpCertCounters`), so a Mithril restore is the one
+write path into `pool_opcert_sequence` whose counters were never checked
+against a chain rule first. `Store.UpdatePoolOpCertSequence` holds the only
+insert into that table, and block application and the Mithril import are its
+only non-test callers, so those two plus the forging pre-flight and the block
+builder are every place the bound has to be named; the rollback recompute
+derives `pool`.`latest_op_cert_sequence` from rows already inside it, and
+`ImportPool` and the pool registration and retirement certificates write that
```
Comment (chrisguiney) [inline ARCHITECTURE.md:4830]: Non-blocking; the claim checks out. `models.Pool.LatestOpCertSequence` has no assignment anywhere except `database/plugin/metadata/sqlstore/pool.go:2305`, which hydrates the model from the column on read, so `ImportPool` always binds zero. Worth recording that the invariant rests on the field never being set rather than on a check: `ImportPool`'s INSERT binds it straight to the signed column with no `checkedInt64`, unlike `UpdatePoolOpCertSequence`. If it ever becomes assignable from snapshot data, that is the one write into the column where the bound would not be named.

**Review by chrisguiney — APPROVED** [review]

File: DATABASE.md [hunk 3]
```diff
@@ -955,7 +955,7 @@ again.
...
... above the tip — so `(pool_key_hash, sequence)` exists to serve it from an index alone: SQLite and PostgreSQL fold it without reading a table row, and MySQL can skip through the index a pool at a time. It is declared by migration `v1alpha1`. The cost is one further index maintained per minted block. `sequence` is a signed engine integer, so the recorded counter domain is `[0, 2^63-1]` even though the reference decodes the operational certificate counter as an unbounded `Word64`. That column carries the monotonicity ordering as well as the value -- the `MAX(sequence)` reads above, the `latest_op_cert_sequence < ?` guard on the denormalized `pool` maximum, and the `(pool_key_hash, sequence)` index -- so a counter above the domain would order below every smaller one. `eras.MaxPersistableOpCertCounter` names the bound; block application refuses a higher counter before applying the block, and the Mithril restore path (`ledgerstate.importOpCertCounters`) refuses it before writing the certified baseline, rather than failing this write part-way through. Every row written before that bound was named is inside it: `checkedInt64` has always refused a higher value, so no row needs rewriting. |
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply DATABASE.md]: Confirmed still present at head a485f49f, and valid as wording — but this clause is pre-existing prose, not authored by this PR. The `pool_opcert_sequence` row is a single line, so the whole cell shows as changed while the only edit is the later sentence naming `ledgerstate.importOpCertCounters`.

If it is corrected, `ledger/eras/opcert.go:67` already has the precise phrasing to match: "The reference imposes no bound: cardano-ledger decodes the counter as Word64 (Cardano.Protocol.TPraos.OCert) and the CDDL declares it uint .size 8". The intended meaning is that the reference imposes no bound narrower than the type, not that `Word64` is infinite.

Non-blocking; not a merge blocker.

[the remark this answers is not in the document]
Comment (chrisguiney) [reply DATABASE.md]: Same finding as CodeRabbit on this line; replied there. Valid wording, but pre-existing prose rather than text this PR authored — the table row is one line, so the entire cell appears in the diff while the only edit is the later sentence about `ledgerstate.importOpCertCounters`. Non-blocking.

File: ledger/eras/opcert_persistable_test.go [hunk 4]
```diff
@@ -70,16 +70,13 @@ func TestValidateOpCertPersistableCounter(t *testing.T) {
...
-// which is what checkedInt64 enforces at the metadata store.
-func TestMaxPersistableOpCertCounterMatchesSignedColumn(t *testing.T) {
-	assert.Equal(
-		t,
-		uint64(math.MaxInt64),
-		eras.MaxPersistableOpCertCounter,
-	)
-}
+// The bound is the largest value a signed 64-bit SQL column holds, which is
+// what sqlstore.checkedInt64 enforces at the metadata store. That coupling is
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply ledger/eras/opcert_persistable_test.go:74]: Valid, and I verified both halves of it against head a485f49f.

You are right that the deleted assertion would have caught a change to the constant: `assert.Equal(t, uint64(math.MaxInt64), eras.MaxPersistableOpCertCounter)` compares against the `math.MaxInt64` literal, so narrowing the constant fails it. The comment’s "would stay green if either side moved" is wrong on that half.

What it could not see is store-side drift, and that half checks out by experiment: narrowing `checkedInt64` to `math.MaxInt32` leaves the whole `ledger/eras` suite green — the package has zero `sqlstore` entries in `go list -deps` — while `ledgerstate.TestPersistableOpCertBoundMatchesStore` fails on the first `UpdatePoolOpCertSequence`. So the swap itself is justified; only the sentence explaining it overstates the gap.

Note the same wording is in the new file too, `ledgerstate/opcert_persistable_bound_test.go` ("it stays green if either side moves"), so a correction should cover both.

Documentation accuracy in a comment, non-blocking — not a merge blocker.

**Review by chrisguiney — DISMISSED** [review]
Verified independently at head a485f49f in a clean worktree.

**Behavior.** The guard is boundary-preserving: `ValidateOpCertPersistableCounter` tests `candidate > MaxPersistableOpCertCounter` and `checkedInt64` tests `value > math.MaxInt64` — same value, same comparison — so the import now refuses exactly what the store already refused, earlier and with the bound named. Nothing consumes the message it replaces: `"exceeds int64"` occurs only at its definition in `sqlstore/operational.go:1255` and in the new negative assertion. `importTip`'s `defer txn.Release()` calls `Rollback()`, so a refused counter leaves no partial rows, as the description claims.

**Class audit — complete, and the description's table matches the code.** `Store.UpdatePoolOpCertSequence` is the only INSERT into `pool_opcert_sequence`, with two non-test callers: `ledger/state.go:7200` (guarded at 6767) and `ledgerstate/import.go:2455` (guarded here). `ValidateOpCertPersistableCounter` now has four non-test call sites — `ledger/state.go:6767`, `ledger/forging/builder.go:833`, `ledger/forging/keys.go:1167`, `ledgerstate/import.go:2448`. For `pool.latest_op_cert_sequence`: the same UPDATE, the rollback recompute from `MAX(sequence)` over already-bounded rows, literal `0` in both certificate inserts, and `ImportPool` binding `models.Pool.LatestOpCertSequence`, which has no assignment anywhere outside the read hydration at `sqlstore/pool.go:2305`. Worth recording alongside that: every read of the column is a plain `uint64(...)` cast with no `checkedUint64` (`sqlstore/pool.go:410`, `434`, `459`, `503`, `2305`), so those reads rest entirely on the write-side invariant this PR completes.

**Tests.** `TestImportOpCertCountersRefusesUnpersistableCounter` is a genuine regression test — reverting the guard in place (keeping the `eras` import referenced so the package still builds) fails it with `"storing certified opcert counter: unsigned SQL value 9223372036854775808 exceeds int64" does not contain "9223372036854775807"`. `TestImportOpCertCountersStoresCounterAtBound` and `TestPersistableOpCertBoundMatchesStore` both pass with the guard reverted; they are guards rather than regression tests, which is the correct role for them. The `ledger/eras` swap is justified on its substance: narrowing `checkedInt64` to `math.MaxInt32` leaves the whole `ledger/eras` suite green (the package has no `sqlstore` in `go list -deps`) and fails `TestPersistableOpCertBoundMatchesStore`.

**Ran at this head.** `go build ./...` and `go vet ./...` clean; `golangci-lint run ./ledgerstate/... ./ledger/eras/...` 0 issues with an isolated cache; `go test -count=1` ok for `./ledgerstate/...`, `./ledger/eras/...`, `./internal/architecture/...`. `gofmt -l` reports `bark/blob_test.go` and `ledger/read_chain_pipeline_test.go`, both already unformatted on `origin/main` — base-branch noise, not this change. All 13 GitHub checks are green, and the merge base equals `origin/main`, so there is no merged-tree gap.

**Not run:** a Mithril restore against a real snapshot, and the conformance/Antithesis suites — no snapshot or infrastructure available in this environment.

The two open bot findings are documentation-accuracy only and non-blocking; replied on both threads. #4018 exists and describes exactly this change.

**Review by arepala-uml — APPROVED** [review]
